internal/controllers/clusterrequest_controller.go

```diff
@@ -805,35 +832,115 @@ func (t *clusterRequestReconcilerTask) handleClusterPolicyConfiguration(ctx cont
+	if configurationAppliedCondition == nil {
+		t.object.Status.ClusterInstanceRef.NonCompliantAt = metav1.Now()
+		return policyTimedOut
+	}
+
+	// If the current status of the Condition is false.
+	if configurationAppliedCondition.Status == metav1.ConditionFalse {
+		switch configurationAppliedCondition.Reason {
+		case string(utils.CRconditionReasons.InProgress):
+			// Check if the configuration application has timed out.
+			if time.Since(t.object.Status.ClusterInstanceRef.NonCompliantAt.Time) > time.Duration(t.object.Spec.Timeout.Configuration) {
+				policyTimedOut = true
+			}
+		case string(utils.CRconditionReasons.TimedOut):
+			policyTimedOut = true
+		case string(utils.CRconditionReasons.Missing):
+			t.object.Status.ClusterInstanceRef.NonCompliantAt = metav1.Now()
+		case string(utils.CRconditionReasons.OutOfDate):
+			t.object.Status.ClusterInstanceRef.NonCompliantAt = metav1.Now()
+		default:
+			t.logger.InfoContext(ctx,
+				fmt.Sprintf("Unexpected Reason for condition type %s",
+					utils.CRconditionTypes.ConfigurationApplied,
+				),
+			)
+		}
+	} else if configurationAppliedCondition.Reason == string(utils.CRconditionReasons.Completed) {
+		t.object.Status.ClusterInstanceRef.NonCompliantAt = metav1.Now()
+	}
+
+	return policyTimedOut
+}
+
+// updateConfigurationAppliedStatus updates the ClusterRequest ConfigurationApplied condition
+// based on the state of the policies matched with the managed cluster.
+func (t *clusterRequestReconcilerTask) updateConfigurationAppliedStatus(
+	ctx context.Context, targetPolicies []oranv1alpha1.PolicyDetails, allPoliciesCompliant bool,
+	nonCompliantPolicyInEnforce bool) error {
+	if len(targetPolicies) == 0 {
+		t.object.Status.ClusterInstanceRef.NonCompliantAt = metav1.Time{}
 		utils.SetStatusCondition(&t.object.Status.Conditions,
 			utils.CRconditionTypes.ConfigurationApplied,
-			utils.CRconditionReasons.Completed,
-			metav1.ConditionTrue,
-			"The configuration is up to date",
+			utils.CRconditionReasons.Missing,
+			metav1.ConditionFalse,
+			"No configuration present",
 		)
 	} else {
-		if nonCompliantPolicyInEnforce {
+		// Update the ConfigurationApplied condition.
+		if allPoliciesCompliant {
+			t.object.Status.ClusterInstanceRef.NonCompliantAt = metav1.Time{}
 			utils.SetStatusCondition(&t.object.Status.Conditions,
 				utils.CRconditionTypes.ConfigurationApplied,
-				utils.CRconditionReasons.InProgress,
-				metav1.ConditionFalse,
-				"The configuration is still being applied",
+				utils.CRconditionReasons.Completed,
+				metav1.ConditionTrue,
+				"The configuration is up to date",
 			)
 		} else {
-			utils.SetStatusCondition(&t.object.Status.Conditions,
-				utils.CRconditionTypes.ConfigurationApplied,
-				utils.CRconditionReasons.OutOfDate,
-				metav1.ConditionFalse,
-				"The configuration is out of date",
-			)
+			if nonCompliantPolicyInEnforce {
+				policyTimedOut := t.hasPolicyConfigurationTimedOut(ctx)
```

Comment:
Why not move the clusterIsReadyForPolicyConfig call/condition here? this way we could list the policies that are matched to this cluster instance before they are getting enforced.

Reply:
Because there are no child policies created if the cluster is not yet ready, so `targetPolicies` would always be empty.